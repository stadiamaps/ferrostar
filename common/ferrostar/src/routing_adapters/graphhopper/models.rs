use serde::{Deserialize};
use alloc::collections::BTreeMap as HashMap;
use crate::models::{ GeographicCoordinate };

#[derive(Deserialize, Debug)]
pub struct GraphHopperRouteResponse {
    pub paths: Vec<GraphHopperPath>,
    pub message: Option<String>
}

#[derive(Deserialize, Debug)]
#[serde(untagged)]
pub enum DetailEntryValue {
    Int(i32),
    Str(String),
    Bool(bool),
    Float(f64),
}

#[derive(Deserialize, Debug)]
#[serde(from = "Vec<serde_json::Value>")]
pub struct DetailEntry {
    pub start_index: usize,
    pub end_index: usize,
    pub value: Option<DetailEntryValue>,
}

impl From<Vec<serde_json::Value>> for DetailEntry {
    fn from(vec: Vec<serde_json::Value>) -> Self {
        let start_index = vec[0].as_i64().unwrap_or_default() as usize;
        let end_index = vec[1].as_i64().unwrap_or_default() as usize;

        let value = match &vec[2] {
            serde_json::Value::String(s) => Some(DetailEntryValue::Str(s.clone())),
            serde_json::Value::Number(n) => {
                if n.is_i64() {
                    Some(DetailEntryValue::Int(n.as_i64().unwrap() as i32))
                } else {
                    Some(DetailEntryValue::Float(n.as_f64().unwrap()))
                }
            },
            serde_json::Value::Bool(b) => Some(DetailEntryValue::Bool(*b)),
            serde_json::Value::Null => None,
            _ => None,
        };

        DetailEntry { start_index, end_index, value }
    }
}

#[derive(Deserialize, Debug)]
pub struct GraphHopperPath {
    pub distance: f64,
    pub time: f64,
    pub bbox: Vec<f64>,
    pub instructions: Vec<GraphHopperInstruction>,
    pub details: HashMap<String, Vec<DetailEntry>>,
    pub points: String, // encoded polyline
    pub points_encoded:	bool,
    pub points_encoded_multiplier: f64,
}

#[derive(Deserialize, Debug)]
pub struct GraphHopperInstruction {
    pub distance: f64,
    pub time: f64,
    pub heading: Option<f64>,
    pub exit_number: Option<u32>,
    pub turn_angle: Option<f64>,
    pub sign: i32,
    pub interval: Vec<usize>,
    pub text: String,
    pub street_ref: Option<String>,
    pub street_name: String,
}

// temporary structure until we understand internal maxspeed handling
pub struct MaxSpeedEntry {
    pub geometry: Vec<GeographicCoordinate>,
    pub speed_limit: Option<f64>,
    pub unit: String,
}
