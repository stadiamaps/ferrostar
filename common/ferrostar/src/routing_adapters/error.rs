#[cfg(feature = "alloc")]
use alloc::string::{String, ToString};

// TODO: This implementation seems less than ideal. In particular, it hides what sort of JSON error occurred due to an apparent bug in UniFFI.
// The trouble appears to be with generating "flat" enum bindings that are used with callback
// interfaces when the underlying actually has fields.
#[derive(Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Error))]
#[cfg_attr(feature = "std", derive(thiserror::Error))]
pub enum InstantiationError {
    #[cfg_attr(
        feature = "std",
        error("Error parsing the JSON options for the request.")
    )]
    OptionsJsonParseError,
}

// TODO: See comment above
#[derive(Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Error))]
#[cfg_attr(feature = "std", derive(thiserror::Error))]
pub enum RoutingRequestGenerationError {
    #[cfg_attr(
        feature = "std",
        error("Too few waypoints were provided to compute a route.")
    )]
    NotEnoughWaypoints,
    #[cfg_attr(feature = "std", error("Error generating JSON for the request."))]
    JsonError,
    #[cfg_attr(
        feature = "std",
        error("An unknown error generating a request was raised in foreign code.")
    )]
    UnknownRequestGenerationError,
}

#[cfg(feature = "uniffi")]
impl From<uniffi::UnexpectedUniFFICallbackError> for RoutingRequestGenerationError {
    fn from(_: uniffi::UnexpectedUniFFICallbackError) -> RoutingRequestGenerationError {
        RoutingRequestGenerationError::UnknownRequestGenerationError
    }
}

impl From<serde_json::Error> for InstantiationError {
    fn from(_: serde_json::Error) -> Self {
        InstantiationError::OptionsJsonParseError
    }
}

impl From<serde_json::Error> for RoutingRequestGenerationError {
    fn from(_: serde_json::Error) -> Self {
        RoutingRequestGenerationError::JsonError
    }
}

#[derive(Debug)]
#[cfg_attr(feature = "std", derive(thiserror::Error))]
#[cfg_attr(feature = "uniffi", derive(uniffi::Error))]
pub enum ParsingError {
    // TODO: Unable to find route and other common errors
    #[cfg_attr(feature = "std", error("Failed to parse route json object: {error}."))]
    InvalidRouteObject { error: String },
    #[cfg_attr(feature = "std", error("Failed to parse route geometry: {error}."))]
    InvalidGeometry { error: String },
    #[cfg_attr(feature = "std", error("Failed to parse annotations: {error}."))]
    MalformedAnnotations { error: String },
    #[cfg_attr(
        feature = "std",
        error("Routing adapter returned an unexpected status code: {code}.")
    )]
    InvalidStatusCode { code: String },
    #[cfg_attr(
        feature = "std",
        error("An unknown error parsing a response was raised in foreign code.")
    )]
    UnknownParsingError,
}

#[cfg(feature = "uniffi")]
impl From<uniffi::UnexpectedUniFFICallbackError> for ParsingError {
    fn from(_: uniffi::UnexpectedUniFFICallbackError) -> ParsingError {
        ParsingError::UnknownParsingError
    }
}

impl From<serde_json::Error> for ParsingError {
    fn from(e: serde_json::Error) -> Self {
        ParsingError::InvalidRouteObject {
            error: e.to_string(),
        }
    }
}
