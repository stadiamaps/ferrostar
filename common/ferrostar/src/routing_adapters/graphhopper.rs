//! High-level HTTP request generation for GraphHopper-based HTTP APIs.

use crate::models::{UserLocation, Waypoint};

use crate::routing_adapters::error::{InstantiationError, RoutingRequestGenerationError};
use crate::routing_adapters::{RouteRequest, RouteRequestGenerator};

use serde::{Deserialize, Serialize};
use serde_json::{json, Map, Value as JsonValue};
#[cfg(feature = "std")]
use std::collections::HashMap;
#[cfg(feature = "wasm-bindgen")]
use tsify::Tsify;

#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[cfg_attr(feature = "wasm-bindgen", derive(Tsify))]
#[cfg_attr(feature = "wasm-bindgen", tsify(from_wasm_abi))]
#[serde(rename_all = "lowercase")]
pub enum GraphHopperVoiceUnits {
    Metric,
    Imperial,
}

/// A route request generator for GraphHopper backends operating over HTTP.
///
/// ## [`WaypointKind`](crate::models::WaypointKind)
///
/// The waypoint kind field of [`Waypoint`] is not currently supported in GraphHopper,
/// as there is no multi-leg support.
/// However, you can configure whether U-turns are allowed with the `pass_through` parameter.
/// Refer to the GraphHopper documentation: <https://docs.graphhopper.com/openapi/routing/getroute>.
///
/// ## Waypoint properties
///
/// The [`Waypoint`] `properties` field is ignored by this route request generator.
///
/// # Examples
///
/// ```
/// use serde_json::{json, Map, Value};
/// use ferrostar::routing_adapters::graphhopper::{GraphHopperHttpRequestGenerator, GraphHopperVoiceUnits};
/// let options: Map<String, Value> = json!({
///     "ch.disable": true,
///     "custom_model": {
///         "distance_influence": 15,
///         "speed": [
///             {
///                 "if": "road_class == MOTORWAY",
///                 "limit_to": "100"
///             }
///         ]
///     }
/// }).as_object().unwrap().to_owned();
/// let request_generator = GraphHopperHttpRequestGenerator::new(
///     "https://graphhopper.com/api/1/navigate/?key=YOUR-API-KEY",
///     "car",
///     "en",
///     GraphHopperVoiceUnits::Metric,
///     options
/// );
/// ```
#[derive(Debug)]
pub struct GraphHopperHttpRequestGenerator {
    endpoint_url: String,
    profile: String,
    locale: String,
    voice_units: GraphHopperVoiceUnits,
    options: Map<String, JsonValue>,
}

impl GraphHopperHttpRequestGenerator {
    /// Creates a new GraphHopper request generator given an endpoint URL, a profile name,
    /// and options to include in the request JSON.
    ///
    /// # Examples
    ///
    /// ```
    /// use serde_json::{json, Map, Value};
    /// use ferrostar::routing_adapters::graphhopper::{GraphHopperHttpRequestGenerator, GraphHopperVoiceUnits};
    /// let options: Map<String, Value> = json!({
    ///     "ch.disable": true,
    ///     "custom_model": {
    ///         "distance_influence": 15,
    ///         "speed": [
    ///             {
    ///                 "if": "road_class == MOTORWAY",
    ///                 "limit_to": "100"
    ///             }
    ///         ]
    ///     }
    /// }).as_object().unwrap().to_owned();
    /// let request_generator = GraphHopperHttpRequestGenerator::new(
    ///     "https://graphhopper.com/api/1/navigate/?key=YOUR-API-KEY",
    ///     "car",
    ///     "en",
    ///     GraphHopperVoiceUnits::Metric,
    ///     options
    /// );
    /// ```
    pub fn new<U: Into<String>, P: Into<String>, L: Into<String>>(
        endpoint_url: U,
        profile: P,
        locale: L,
        voice_units: GraphHopperVoiceUnits,
        options: Map<String, JsonValue>,
    ) -> Self {
        Self {
            endpoint_url: endpoint_url.into(),
            profile: profile.into(),
            locale: locale.into(),
            voice_units,
            options,
        }
    }

    /// Creates a new GraphHopper request generator given an endpoint URL, a profile name,
    /// and options to include in the request JSON.
    /// Options in this constructor are a JSON fragment representing any
    /// options you want to add along with the request.
    ///
    /// # Examples
    ///
    /// ```
    /// # use ferrostar::routing_adapters::graphhopper::{GraphHopperHttpRequestGenerator, GraphHopperVoiceUnits};
    /// let options = r#"{
    ///     "ch.disable": true,
    ///     "custom_model": {
    ///         "distance_influence": 15,
    ///         "speed": [
    ///             {
    ///                 "if": "road_class == MOTORWAY",
    ///                 "limit_to": "100"
    ///             }
    ///         ]
    ///     }
    /// }"#;
    ///
    /// // Without options
    /// let request_generator = GraphHopperHttpRequestGenerator::with_options_json(
    ///     "https://graphhopper.com/api/1/navigate/?key=YOUR-API-KEY",
    ///     "car",
    ///     "en",
    ///     GraphHopperVoiceUnits::Metric,
    ///     None
    /// );
    ///
    /// // With options
    /// let request_generator = GraphHopperHttpRequestGenerator::with_options_json(
    ///     "https://graphhopper.com/api/1/navigate/?key=YOUR-API-KEY",
    ///     "car",
    ///     "en",
    ///     GraphHopperVoiceUnits::Metric,
    ///     Some(options)
    /// );
    /// ```
    pub fn with_options_json<U: Into<String>, P: Into<String>, L: Into<String>>(
        endpoint_url: U,
        profile: P,
        locale: L,
        voice_units: GraphHopperVoiceUnits,
        options_json: Option<&str>,
    ) -> Result<Self, InstantiationError> {
        let parsed_options = match options_json {
            Some(options) => serde_json::from_str::<JsonValue>(options)?
                .as_object()
                .ok_or(InstantiationError::OptionsJsonParseError)?
                .to_owned(),
            None => Map::new(),
        };
        Ok(Self {
            endpoint_url: endpoint_url.into(),
            profile: profile.into(),
            locale: locale.into(),
            voice_units,
            options: parsed_options,
        })
    }
}

impl RouteRequestGenerator for GraphHopperHttpRequestGenerator {
    fn generate_request(
        &self,
        user_location: UserLocation,
        waypoints: Vec<Waypoint>,
    ) -> Result<RouteRequest, RoutingRequestGenerationError> {
        if waypoints.is_empty() {
            Err(RoutingRequestGenerationError::NotEnoughWaypoints)
        } else {
            let headers =
                HashMap::from([("Content-Type".to_string(), "application/json".to_string())]);

            let mut points: Vec<Vec<f64>> = vec![vec![
                user_location.coordinates.lng,
                user_location.coordinates.lat,
            ]];

            points.extend(
                waypoints
                    .iter()
                    .map(|waypoint| vec![waypoint.coordinate.lng, waypoint.coordinate.lat]),
            );

            let mut args = json!({
                "profile": &self.profile,
                "points": points,
                "locale": self.locale,
                "type": "mapbox",
                "voice_units": self.voice_units,
            });

            for (k, v) in &self.options {
                args[k] = v.clone();
            }

            let body = serde_json::to_vec(&args)?;
            Ok(RouteRequest::HttpPost {
                url: self.endpoint_url.clone(),
                headers,
                body,
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::{GeographicCoordinate, WaypointKind};
    use serde_json::from_slice;

    #[cfg(all(feature = "std", not(feature = "web-time")))]
    use std::time::SystemTime;

    #[cfg(feature = "web-time")]
    use web_time::SystemTime;

    const ENDPOINT_URL: &str = "https://graphhopper.com/api/1/navigate/?key=YOUR-API-KEY";
    const COSTING: &str = "car";
    const LOCALE: &str = "en";
    const VOICE_UNITS: GraphHopperVoiceUnits = GraphHopperVoiceUnits::Metric;

    const USER_LOCATION: UserLocation = UserLocation {
        coordinates: GeographicCoordinate { lat: 0.0, lng: 0.0 },
        horizontal_accuracy: 6.0,
        course_over_ground: None,
        timestamp: SystemTime::UNIX_EPOCH,
        speed: None,
    };
    const WAYPOINTS: [Waypoint; 2] = [
        Waypoint {
            coordinate: GeographicCoordinate { lat: 0.0, lng: 1.0 },
            kind: WaypointKind::Break,
            properties: None,
        },
        Waypoint {
            coordinate: GeographicCoordinate { lat: 2.0, lng: 3.0 },
            kind: WaypointKind::Break,
            properties: None,
        },
    ];

    #[test]
    fn not_enough_locations() {
        let generator = GraphHopperHttpRequestGenerator::new(
            ENDPOINT_URL,
            COSTING,
            LOCALE,
            VOICE_UNITS,
            Map::new(),
        );

        // At least two locations are required
        assert!(matches!(
            generator.generate_request(USER_LOCATION, Vec::new()),
            Err(RoutingRequestGenerationError::NotEnoughWaypoints)
        ));
    }

    fn generate_body(
        user_location: UserLocation,
        waypoints: Vec<Waypoint>,
        options_json: Option<&str>,
    ) -> JsonValue {
        let generator = GraphHopperHttpRequestGenerator::with_options_json(
            ENDPOINT_URL,
            COSTING,
            LOCALE,
            VOICE_UNITS,
            options_json,
        )
        .expect("Unable to create request generator");

        match generator.generate_request(user_location, waypoints) {
            Ok(RouteRequest::HttpPost {
                url: request_url,
                headers,
                body,
            }) => {
                assert_eq!(ENDPOINT_URL, request_url);
                assert_eq!(headers["Content-Type"], "application/json".to_string());
                from_slice(&body).expect("Failed to parse request body as JSON")
            }
            Ok(RouteRequest::HttpGet { .. }) => unreachable!(
                "The GraphHopper HTTP request generator currently only generates POST requests"
            ),
            Err(e) => {
                println!("Failed to generate request: {:?}", e);
                json!(null)
            }
        }
    }

    #[test]
    fn request_body_without_options() {
        insta::assert_json_snapshot!(generate_body(USER_LOCATION, WAYPOINTS.to_vec(), None))
    }

    #[test]
    fn request_body_with_custom_profile() {
        insta::assert_json_snapshot!(generate_body(
            USER_LOCATION,
            WAYPOINTS.to_vec(),
            Some(
                r#"{
                    "ch.disable": true,
                    "custom_model": {
                        "distance_influence": 15,
                        "speed": [
                            {
                                "if": "road_class == MOTORWAY",
                                "limit_to": "100"
                            }
                        ]
                    }
                }"#
            )
        ))
    }
}
