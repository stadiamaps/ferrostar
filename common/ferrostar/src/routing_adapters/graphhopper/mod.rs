use crate::models::{UserLocation, Waypoint};

use crate::routing_adapters::error::{InstantiationError, RoutingRequestGenerationError};
use crate::routing_adapters::{RouteRequest, RouteRequestGenerator};

use serde_json::{json, Map, Value as JsonValue};
#[cfg(feature = "std")]
use std::collections::HashMap;

#[derive(Debug)]
pub struct GraphHopperHttpRequestGenerator {
    endpoint_url: String,
    profile: String,
    options: Map<String, JsonValue>,
}

impl GraphHopperHttpRequestGenerator {
    pub fn new(endpoint_url: String, profile: String, options: Map<String, JsonValue>) -> Self {
        Self {
            endpoint_url,
            profile,
            options,
        }
    }

    pub fn with_options_json(
        endpoint_url: String,
        profile: String,
        options_json: Option<&str>,
    ) -> Result<Self, InstantiationError> {
        let parsed_options = match options_json {
            // TODO: Another error variant
            Some(options) => serde_json::from_str::<JsonValue>(options)?
                .as_object()
                .ok_or(InstantiationError::OptionsJsonParseError)?
                .to_owned(),
            None => Map::new(),
        };
        Ok(Self {
            endpoint_url,
            profile,
            options: parsed_options,
        })
    }
}

impl RouteRequestGenerator for GraphHopperHttpRequestGenerator {
    fn generate_request(
        &self,
        user_location: UserLocation,
        waypoints: Vec<Waypoint>,
    ) -> Result<RouteRequest, RoutingRequestGenerationError> {
        if waypoints.is_empty() {
            Err(RoutingRequestGenerationError::NotEnoughWaypoints)
        } else {
            let headers =
                HashMap::from([("Content-Type".to_string(), "application/json".to_string())]);

            let mut points: Vec<Vec<f64>> = vec![vec![
                user_location.coordinates.lng,
                user_location.coordinates.lat,
            ]];

            points.extend(
                waypoints
                    .iter()
                    .map(|waypoint| vec![waypoint.coordinate.lng, waypoint.coordinate.lat]),
            );

            let mut args = json!({
                "profile": &self.profile,
                "points": points,
                "locale": "en",
                "type": "mapbox",
                "voice_units": "metric",
            });

            for (k, v) in &self.options {
                args[k] = v.clone();
            }

            let body = serde_json::to_vec(&args)?;
            Ok(RouteRequest::HttpPost {
                url: self.endpoint_url.clone(),
                headers,
                body,
            })
        }
    }
}
