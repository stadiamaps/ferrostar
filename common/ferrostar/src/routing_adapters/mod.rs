//! The bridge between routing engines and Ferrostar in high-level platform frameworks.
//!
//! This module provides:
//! - Important traits for framework implementers
//! - Generic HTTP request generation and response parsing for common routing APIs
//!
//! If you're reading this module documentation,
//! you are probably bringing Ferrostar to a new platform
//! where no high-level library like the native iOS and Android `FerrostarCore` exists.
//!
//! The first thing you should know is that
//! there isn't anything inherently special about the types at the root of this module.
//! These types are designed to be integral to higher-level platform libraries
//! (like those for iOS and Android) to ensure extensibility.
//! But they aren't used anywhere else in the crate.
//!
//! If you're implementing Ferrostar bindings for a new platform,
//! have a look at the [iOS implementation](https://github.com/stadiamaps/ferrostar/blob/main/apple/Sources/FerrostarCore/FerrostarCore.swift),
//! particularly how the initializer requires a [`RouteProvider`](https://github.com/stadiamaps/ferrostar/blob/main/apple/Sources/FerrostarCore/RouteProvider.swift).
//! All routing in the `FerrostarCore` class goes through this,
//! with the heavy lifting done by either a [`RouteAdapter`] or a custom provider
//! (native arbitrary code).
//! We suggest that other platforms follow a similar approach to maximize extensibility.
//!
//! If you're doing a direct integration calling this library directly,
//! you can probably go straight to the submodules for integrations with major routing engines
//! and then construct a [`NavigationController`](crate::navigation_controller::NavigationController)
//! with a route and configuration.
//! PRs are welcome for any routing API with an open specification (even if it's a commercial offering).
//!
//! Ferrostar also fully supports proprietary and on-device routing.
//! All you need to do is convert your routes into Ferrostar [Route]s,
//! and nothing in this module is strictly required to do that.

use crate::models::Waypoint;
use crate::models::{Route, UserLocation};
use crate::routing_adapters::error::InstantiationError;
use error::{ParsingError, RoutingRequestGenerationError};

#[cfg(all(not(feature = "std"), feature = "alloc"))]
use alloc::collections::BTreeMap as HashMap;
#[cfg(feature = "std")]
use std::{collections::HashMap, fmt::Debug};

#[cfg(feature = "wasm-bindgen")]
use serde_json::json;
#[cfg(feature = "wasm-bindgen")]
use wasm_bindgen::{prelude::wasm_bindgen, JsValue};

#[cfg(feature = "alloc")]
use alloc::{string::String, sync::Arc, vec::Vec};

use crate::routing_adapters::osrm::OsrmResponseParser;
use crate::routing_adapters::valhalla::ValhallaHttpRequestGenerator;

pub mod error;
pub mod osrm;
pub mod utilities;
pub mod valhalla;

/// A route request generated by a [`RouteRequestGenerator`].
#[derive(PartialEq, Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
pub enum RouteRequest {
    HttpPost {
        url: String,
        headers: HashMap<String, String>,
        body: Vec<u8>,
    },
    HttpGet {
        url: String,
        headers: HashMap<String, String>,
    },
}

/// A trait describing any object capable of generating [`RouteRequest`]s.
///
/// The interface is intentionally generic. Every routing backend has its own set of
/// parameters, including a "profile," max travel speed, units of speed and distance, and more.
/// It is assumed that these properties will be set at construction time or otherwise configured
/// before use, so that we can keep the public interface as generic as possible.
///
/// Implementations may be either in Rust (most popular engines should eventually have Rust
/// glue code) or foreign code.
#[cfg_attr(feature = "uniffi", uniffi::export(with_foreign))]
pub trait RouteRequestGenerator: Send + Sync {
    /// Generates a routing backend request given the set of locations.
    ///
    /// While most implementations will treat the locations as an ordered sequence, this is not
    /// guaranteed (ex: an optimized router).
    // TODO: Arbitrary options; how can we make this generic???
    // TODO: Option for whether we should account for course over ground or heading.
    fn generate_request(
        &self,
        user_location: UserLocation,
        waypoints: Vec<Waypoint>,
    ) -> Result<RouteRequest, RoutingRequestGenerationError>;

    // TODO: "Trace attributes" request method? Maybe in a separate trait?
}

/// A generic interface describing any object capable of parsing a response from a routing
/// backend into one or more [`Route`]s.
#[cfg_attr(feature = "uniffi", uniffi::export(with_foreign))]
pub trait RouteResponseParser: Send + Sync {
    /// Parses a raw response from the routing backend into a route.
    ///
    /// We use a sequence of octets as a common interchange format.
    /// as this works for all currently conceivable formats (JSON, PBF, etc.).
    fn parse_response(&self, response: Vec<u8>) -> Result<Vec<Route>, ParsingError>;
}

/// The route adapter bridges between the common core and a routing backend where interaction takes place
/// over a generic request/response flow (typically over a network;
/// local/offline routers **do not use this object** as the interaction patterns are different).
///
/// This is essentially the composite of the [`RouteRequestGenerator`] and [`RouteResponseParser`]
/// traits, but it provides one further level of abstraction which is helpful to consumers.
/// As there is no way to signal compatibility between request generators and response parsers,
/// the [`RouteAdapter`] provides convenience constructors which take the guesswork out of it,
/// while still leaving consumers free to implement one or both halves.
///
/// In the future, we may provide additional methods or conveniences, and this
/// indirection leaves the design open to such changes without necessarily breaking source
/// compatibility.
/// One such possible extension would be the ability to fetch more detailed attributes in real time.
/// This is supported by the Valhalla stack, among others.
///
/// Ideas  welcome re: how to signal compatibility between request generators and response parsers.
/// I don't think we can do this in the type system, since one of the reasons for the split design
/// is modularity, including the possibility of user-provided implementations, and these will not
/// always be of a "known" type to the Rust side.
#[cfg_attr(feature = "uniffi", derive(uniffi::Object))]
pub struct RouteAdapter {
    request_generator: Arc<dyn RouteRequestGenerator>,
    response_parser: Arc<dyn RouteResponseParser>,
}

#[cfg_attr(feature = "uniffi", uniffi::export)]
impl RouteAdapter {
    #[cfg_attr(feature = "uniffi", uniffi::constructor)]
    pub fn new(
        request_generator: Arc<dyn RouteRequestGenerator>,
        response_parser: Arc<dyn RouteResponseParser>,
    ) -> Self {
        Self {
            request_generator,
            response_parser,
        }
    }

    #[cfg_attr(feature = "uniffi", uniffi::constructor)]
    pub fn new_valhalla_http(
        endpoint_url: String,
        profile: String,
        options_json: Option<String>,
    ) -> Result<Self, InstantiationError> {
        let request_generator = Arc::new(ValhallaHttpRequestGenerator::with_options_json(
            endpoint_url,
            profile,
            options_json.as_deref(),
        )?);
        let response_parser = Arc::new(OsrmResponseParser::new(6));
        Ok(Self::new(request_generator, response_parser))
    }

    //
    // Proxied implementation methods.
    //

    pub fn generate_request(
        &self,
        user_location: UserLocation,
        waypoints: Vec<Waypoint>,
    ) -> Result<RouteRequest, RoutingRequestGenerationError> {
        self.request_generator
            .generate_request(user_location, waypoints)
    }

    pub fn parse_response(&self, response: Vec<u8>) -> Result<Vec<Route>, ParsingError> {
        self.response_parser.parse_response(response)
    }
}

/// JavaScript wrapper for `RouteAdapter`.
#[cfg(feature = "wasm-bindgen")]
#[wasm_bindgen(js_name = RouteAdapter)]
pub struct JsRouteAdapter(RouteAdapter);

#[cfg(feature = "wasm-bindgen")]
#[wasm_bindgen(js_class = RouteAdapter)]
impl JsRouteAdapter {
    /// Creates a new RouteAdapter with a Valhalla HTTP request generator and an OSRM response parser.
    /// At the moment, this is the only supported combination.
    #[wasm_bindgen(constructor)]
    pub fn new(
        endpoint_url: String,
        profile: String,
        costing_options_json: Option<String>,
    ) -> Result<JsRouteAdapter, JsValue> {
        RouteAdapter::new_valhalla_http(endpoint_url, profile, costing_options_json)
            .map(JsRouteAdapter)
            .map_err(|e| JsValue::from_str(&format!("{}", e)))
        // TODO: We should have a better error handling strategy here. Same for the other methods.
    }

    #[wasm_bindgen(js_name = generateRequest)]
    pub fn generate_request(
        &self,
        user_location: JsValue,
        waypoints: JsValue,
    ) -> Result<JsValue, JsValue> {
        let user_location: UserLocation = serde_wasm_bindgen::from_value(user_location)
            .map_err(|e| JsValue::from_str(&e.to_string()))?;

        let waypoints: Vec<Waypoint> = serde_wasm_bindgen::from_value(waypoints)
            .map_err(|e| JsValue::from_str(&e.to_string()))?;

        match self.0.generate_request(user_location, waypoints) {
            Ok(RouteRequest::HttpPost { url, headers, body }) => {
                serde_wasm_bindgen::to_value(&json!({
                    "method": "post",
                    "url": url,
                    "headers": headers,
                    "body": body,
                }))
                .map_err(|e| JsValue::from_str(&e.to_string()))
            }
            Ok(RouteRequest::HttpGet { url, headers }) => serde_wasm_bindgen::to_value(&json!({
                "method": "get",
                "url": url,
                "headers": headers,
            }))
            .map_err(|e| JsValue::from_str(&e.to_string())),
            Err(e) => Err(JsValue::from_str(&e.to_string())),
        }
    }

    #[wasm_bindgen(js_name = parseResponse)]
    pub fn parse_response(&self, response: Vec<u8>) -> Result<JsValue, JsValue> {
        match self.0.parse_response(response.into()) {
            Ok(routes) => serde_wasm_bindgen::to_value(&routes).map_err(JsValue::from),
            Err(error) => Err(JsValue::from_str(&error.to_string())),
        }
    }
}
