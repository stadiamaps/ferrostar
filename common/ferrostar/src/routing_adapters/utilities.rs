use polyline::decode_polyline;

use crate::models::GeographicCoordinate;

use super::error::ParsingError;

/// Parse a polyline-encoded geometry string into a list of geographic coordinates.
/// If the polyline cannot be decoded, a [`ParsingError`] results.
pub fn get_coordinates_from_geometry(
    geometry: &str,
    polyline_precision: u32,
) -> Result<Vec<GeographicCoordinate>, ParsingError> {
    let linestring = decode_polyline(geometry, polyline_precision).map_err(|error| {
        ParsingError::InvalidGeometry {
            error: error.to_string(),
        }
    })?;

    // TODO: Trait for this common pattern?
    let linestring = linestring
        .coords()
        .map(|coord| GeographicCoordinate::from(*coord))
        .collect();

    Ok(linestring)
}
