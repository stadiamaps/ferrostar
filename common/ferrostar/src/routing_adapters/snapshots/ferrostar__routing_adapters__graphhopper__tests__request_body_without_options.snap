---
source: ferrostar/src/routing_adapters/graphhopper.rs
expression: "generate_body(USER_LOCATION, WAYPOINTS.to_vec(), None)"
---
{
  "locale": "en",
  "points": [
    [
      0.0,
      0.0
    ],
    [
      1.0,
      0.0
    ],
    [
      3.0,
      2.0
    ]
  ],
  "profile": "car",
  "type": "mapbox",
  "voice_units": "metric"
}
